// revolt.js/dist/permissions/definitions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_PERMISSION_SERVER = exports.DEFAULT_PERMISSION_DIRECT_MESSAGE = exports.DEFAULT_PERMISSION_SAVED_MESSAGES = exports.DEFAULT_PERMISSION = exports.DEFAULT_PERMISSION_VIEW_ONLY = exports.ALLOW_IN_TIMEOUT = exports.U32_MAX = exports.Permission = exports.UserPermission = void 0;
/**
 * Permission against User
 */
exports.UserPermission = {
    Access: 1 << 0,
    ViewProfile: 1 << 1,
    SendMessage: 1 << 2,
    Invite: 1 << 3,
};
/**
 * Permission against Server / Channel
 */
exports.Permission = {
    // * Generic permissions
    /// Manage the channel or channels on the server
    ManageChannel: Math.pow(2, 0),
    /// Manage the server
    ManageServer: Math.pow(2, 1),
    /// Manage permissions on servers or channels
    ManagePermissions: Math.pow(2, 2),
    /// Manage roles on server
    ManageRole: Math.pow(2, 3),
    /// Manage server customisation (includes emoji)
    ManageCustomisation: Math.pow(2, 4),
    // % 1 bits reserved
    // * Member permissions
    /// Kick other members below their ranking
    KickMembers: Math.pow(2, 6),
    /// Ban other members below their ranking
    BanMembers: Math.pow(2, 7),
    /// Timeout other members below their ranking
    TimeoutMembers: Math.pow(2, 8),
    /// Assign roles to members below their ranking
    AssignRoles: Math.pow(2, 9),
    /// Change own nickname
    ChangeNickname: Math.pow(2, 10),
    /// Change or remove other's nicknames below their ranking
    ManageNicknames: Math.pow(2, 11),
    /// Change own avatar
    ChangeAvatar: Math.pow(2, 12),
    /// Remove other's avatars below their ranking
    RemoveAvatars: Math.pow(2, 13),
    // % 7 bits reserved
    // * Channel permissions
    /// View a channel
    ViewChannel: Math.pow(2, 20),
    /// Read a channel's past message history
    ReadMessageHistory: Math.pow(2, 21),
    /// Send a message in a channel
    SendMessage: Math.pow(2, 22),
    /// Delete messages in a channel
    ManageMessages: Math.pow(2, 23),
    /// Manage webhook entries on a channel
    ManageWebhooks: Math.pow(2, 24),
    /// Create invites to this channel
    InviteOthers: Math.pow(2, 25),
    /// Send embedded content in this channel
    SendEmbeds: Math.pow(2, 26),
    /// Send attachments and media in this channel
    UploadFiles: Math.pow(2, 27),
    /// Masquerade messages using custom nickname and avatar
    Masquerade: Math.pow(2, 28),
    /// React to messages with emoji
    React: Math.pow(2, 29),
    // * Voice permissions
    /// Connect to a voice channel
    Connect: Math.pow(2, 30),
    /// Speak in a voice call
    Speak: Math.pow(2, 31),
    /// Share video in a voice call
    Video: Math.pow(2, 32),
    /// Mute other members with lower ranking in a voice call
    MuteMembers: Math.pow(2, 33),
    /// Deafen other members with lower ranking in a voice call
    DeafenMembers: Math.pow(2, 34),
    /// Move members between voice channels
    MoveMembers: Math.pow(2, 35),
    // * Misc. permissions
    // % Bits 36 to 52: free area
    // % Bits 53 to 64: do not use
    // * Grant all permissions
    /// Safely grant all permissions
    GrantAllSafe: 4503599627370495,
};
/**
 * Maximum safe value
 */
exports.U32_MAX = Math.pow(2, 32) - 1; // 4294967295
/**
 * Permissions allowed for a user while in timeout
 */
exports.ALLOW_IN_TIMEOUT = exports.Permission.ViewChannel + exports.Permission.ReadMessageHistory;
/**
 * Default permissions if we can only view
 */
exports.DEFAULT_PERMISSION_VIEW_ONLY = exports.Permission.ViewChannel + exports.Permission.ReadMessageHistory;
/**
 * Default base permissions for channels
 */
exports.DEFAULT_PERMISSION = exports.DEFAULT_PERMISSION_VIEW_ONLY +
    exports.Permission.SendMessage +
    exports.Permission.InviteOthers +
    exports.Permission.SendEmbeds +
    exports.Permission.UploadFiles +
    exports.Permission.Connect +
    exports.Permission.Speak;
/**
 * Permissions in saved messages channel
 */
exports.DEFAULT_PERMISSION_SAVED_MESSAGES = exports.Permission.GrantAllSafe;
/**
 * Permissions in direct message channels / default permissions for group DMs
 */
exports.DEFAULT_PERMISSION_DIRECT_MESSAGE = exports.DEFAULT_PERMISSION + exports.Permission.React + exports.Permission.ManageChannel;
/**
 * Permissions in server text / voice channel
 */
exports.DEFAULT_PERMISSION_SERVER = exports.DEFAULT_PERMISSION + exports.Permission.React + exports.Permission.ChangeNickname + exports.Permission.ChangeAvatar;
