// revolt.js/dist/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.API = exports.toNullableDate = exports.toNullable = exports.calculatePermission = exports.U32_MAX = exports.DEFAULT_PERMISSION_VIEW_ONLY = exports.DEFAULT_PERMISSION_SERVER = exports.DEFAULT_PERMISSION_SAVED_MESSAGES = exports.DEFAULT_PERMISSION_DIRECT_MESSAGE = exports.DEFAULT_PERMISSION = exports.Permission = exports.UserPermission = exports.Emoji = exports.User = exports.Server = exports.Message = exports.Member = exports.Channel = void 0;
var Channels_1 = require("./maps/Channels");
Object.defineProperty(exports, "Channel", { enumerable: true, get: function () { return Channels_1.Channel; } });
var Members_1 = require("./maps/Members");
Object.defineProperty(exports, "Member", { enumerable: true, get: function () { return Members_1.Member; } });
var Messages_1 = require("./maps/Messages");
Object.defineProperty(exports, "Message", { enumerable: true, get: function () { return Messages_1.Message; } });
var Servers_1 = require("./maps/Servers");
Object.defineProperty(exports, "Server", { enumerable: true, get: function () { return Servers_1.Server; } });
var Users_1 = require("./maps/Users");
Object.defineProperty(exports, "User", { enumerable: true, get: function () { return Users_1.User; } });
var Emojis_1 = require("./maps/Emojis");
Object.defineProperty(exports, "Emoji", { enumerable: true, get: function () { return Emojis_1.Emoji; } });
__exportStar(require("./Client"), exports);
__exportStar(require("./config"), exports);
var definitions_1 = require("./permissions/definitions");
Object.defineProperty(exports, "UserPermission", { enumerable: true, get: function () { return definitions_1.UserPermission; } });
Object.defineProperty(exports, "Permission", { enumerable: true, get: function () { return definitions_1.Permission; } });
Object.defineProperty(exports, "DEFAULT_PERMISSION", { enumerable: true, get: function () { return definitions_1.DEFAULT_PERMISSION; } });
Object.defineProperty(exports, "DEFAULT_PERMISSION_DIRECT_MESSAGE", { enumerable: true, get: function () { return definitions_1.DEFAULT_PERMISSION_DIRECT_MESSAGE; } });
Object.defineProperty(exports, "DEFAULT_PERMISSION_SAVED_MESSAGES", { enumerable: true, get: function () { return definitions_1.DEFAULT_PERMISSION_SAVED_MESSAGES; } });
Object.defineProperty(exports, "DEFAULT_PERMISSION_SERVER", { enumerable: true, get: function () { return definitions_1.DEFAULT_PERMISSION_SERVER; } });
Object.defineProperty(exports, "DEFAULT_PERMISSION_VIEW_ONLY", { enumerable: true, get: function () { return definitions_1.DEFAULT_PERMISSION_VIEW_ONLY; } });
Object.defineProperty(exports, "U32_MAX", { enumerable: true, get: function () { return definitions_1.U32_MAX; } });
var calculator_1 = require("./permissions/calculator");
Object.defineProperty(exports, "calculatePermission", { enumerable: true, get: function () { return calculator_1.calculatePermission; } });
var null_1 = require("./util/null");
Object.defineProperty(exports, "toNullable", { enumerable: true, get: function () { return null_1.toNullable; } });
Object.defineProperty(exports, "toNullableDate", { enumerable: true, get: function () { return null_1.toNullableDate; } });
exports.API = __importStar(require("revolt-api"));
